File: src/main/java/io/github/ajoz/workshop/fp/java/part_2/solutions/exercise_1/Exercise1.java
package io.github.ajoz.workshop.fp.java.part_2.solutions.exercise_1;

import io.github.ajoz.workshop.fp.java.tools.Function2;

import java.util.List;

class Exercise1 {
    public static Integer sum(final List<Integer> list) {
        Integer sum = 0;
        for (Integer value : list) {
            sum += value;
        }

        return sum;
    }

    public static Integer product(final List<Integer> list) {
        Integer product = 1;
        for (Integer value : list) {
            product *= value;
        }

        return product;
    }

    public static Integer foo(final List<Integer> list,
                              final Integer initial,
                              final Function2<Integer, Integer, Integer> operator) {
        Integer accumulator = initial;
        for (final Integer element : list) {
            accumulator = operator.apply(accumulator, element);
        }

        return accumulator;
    }

    // now let's rewrite sum in terms of foo
    public static Integer sum2(final List<Integer> list) {
        return foo(list, 0, (a, b) -> a + b);
    }

    // now let's rewrite product in terms of foo
    public static Integer product2(final List<Integer> list) {
        return foo(list, 1, (a, b) -> a * b);
    }
}

File: src/main/kotlin/io/github/ajoz/workshop/fp/kotlin/part_2/solutions/exercise_1/Exercise1.kt
@file:Suppress("PackageName", "UNUSED_PARAMETER")

package io.github.ajoz.workshop.fp.kotlin.part_2.solutions.exercise_1

fun sum(list: List<Int>): Int {
    var sum = 0
    for (value in list) {
        sum += value
    }
    return sum
}

fun product(list: List<Int>): Int {
    var sum = 1
    for (value in list) {
        sum *= value
    }
    return sum
}


// for the abstracted function we will use a name `foo`
// hint:
// - what arguments do we need to pass to `foo`?
// - `sum` is using + operator, `product` is using * operator
//   how can we abstract these two operators?
// - do not use any conditional statements for this exercise
fun foo(list: List<Int>, initial: Int, operator: (Int, Int) -> Int): Int {
    var accumulator = initial
    for (value in list) {
        accumulator = operator(value, accumulator)
    }
    return accumulator
}

fun sum2(list: List<Int>): Int {
    return foo(list, 0) { a, b ->
        a + b
    }
}

fun product2(list: List<Int>): Int {
    return foo(list, 1) { a, b ->
        a * b
    }
}
